### DISCUSSION TITLE
Why the massive AppDeployToolkitMain.ps1?

### DISCUSSION THREAD

**Discussion by khaffner** [PSAppDeployToolkit/PSAppDeployToolkit#859] [General]
Why not break it up so each function has their own file named after the function it contains? One giant script made more sense in the powershell_ise days, but now with VSCode working with a folder with lots of ps1 files is a breeze. The AppDeployToolkitMain.ps1 can simply dotsource all those files.

I believe this would also be a great step towards making PSADT a proper powershell module. What do you guys think?

I'll be happy to actually do it and create a PR if people want it (and it doesn't break anything of course).

**Comment by sintaxasn** [comment]
Hi Kevin, this is something we've been back and forth about. Modules are typically still compiled into a single monolithic psm1 with individual functions broken out for development purposes only. The reason for this is usually speed, I verified this myself with some rough benchmarking on some early v4 builds. Split out to individual files was taking anywhere up to 13 seconds longer to load that a single monolithic file.

We have other things to factor in here too - like dot-sourcing vs module - and other architecture items that will impact where we settle on this. More to come soon - gotta get some of this stuff properly mapped out to a GitHub project with milestones etc.

**Reply by khaffner** [reply]
Interesting, 13 seconds is a lot.

**Reply by khaffner** [reply]
I just did some testing myself. Took the master branch of psadt, ripped all functions out of AppDeployToolkitMain.ps1 and placed them into their own ps1 files in the same folder, and left this snippet there instead:
![image](https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/assets/13216618/4bd9c690-5c9f-4f4a-8936-45c8ed11b011)

I also set Toolkit_RequireAdmin and Toolkit_LogWriteToHost to False, in both master and my branch for the benchmark. Just to remove factors in the terminal that can slow things down. I ran a dot sourcing of the Main.ps1 10 times, and noted the average.
`1..10 | foreach {(Measure-Command {. .\AppDeployToolkitMain.ps1}).Seconds}`

Results: 
PSADT as of commit 0958724: 6,8 seconds
PSADT with functions in their own files: 5,5 seconds

My fork: https://github.com/khaffner/PSAppDeployToolkit/tree/split-toolkit/Toolkit/AppDeployToolkit
